File: TypeScript/123_Best_Time_to_Buy_and_Sell_Stock_III.ts
function maxProfit(prices: number[]): number {
    const n: number = prices.length;
    if(n<2) return 0;

    let min_price: number = prices[0];
    let max_price: number = prices[n-1];
    const profit1: number[] = new Array(n).fill(0);
    const profit2: number[] = new Array(n).fill(0);

    for(let i=1;i<n;i++){
        profit1[i] = Math.max(profit1[i-1], prices[i] - min_price);
        min_price = Math.min(min_price, prices[i]);

        const j: number = n-i-1;
        profit2[j] = Math.max(profit2[j+1], max_price - prices[j]);
        max_price = Math.max(max_price, prices[j]);
    }
    let ans: number = 0;

    for(let i=0;i<n;i++){
        ans = Math.max(ans, profit1[i] + profit2[i]);
    }

    return ans;
};

File: Python/123_Best_Time_to_Buy_and_Sell_Stock_III.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        # If the lenght is less than 2 there are no posible profits
        if n < 2:
            return 0
        
        # Variables to save min and max price
        min_price = prices[0]
        max_price = prices[n-1]
        # List to save the profit at each position
        # Forward
        profit1 = [0] * n 
        # Backward
        profit2 = [0] * n

        # Save the profits and the min and max prices
        for i in range(1,n):
            profit1[i] = max(profit1[i-1], prices[i] - min_price) # Update the profit at position i if its greater
            min_price = min(min_price, prices[i])  # Update minimum price

            j = n-1-i 
            profit2[j] = max(profit2[j+1], max_price - prices[j]) # Update the profit at position j if its greater
            max_price = max(max_price, prices[j])  # Update maximum price

        ans = 0 

        # Search the maximum profit using the profit1 plus profit2 at position i
        for i in range(n):
            ans = max(ans, profit1[i] + profit2[i])

        return ans
